File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/signing/core/GuardWire.kt
package app.getvela.wallet.feature.signing.core

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/*
 * The `approval_guard` machine's wire (spec 044) — an unlimited approval
 * never leaves the wallet. `chain_id`/`decimals`/`index`/`word_index`/
 * `amount_bits` are `u32`, `now_ms` `f64`.
 */

@Serializable
enum class GuardApprovalKind {
    @SerialName("erc20_approve") Erc20Approve,

    @SerialName("increase_allowance") IncreaseAllowance,

    @SerialName("decrease_allowance") DecreaseAllowance,

    @SerialName("set_approval_for_all") SetApprovalForAll,

    @SerialName("erc2612_permit") Erc2612Permit,

    @SerialName("dai_permit") DaiPermit,

    @SerialName("permit2_single") Permit2Single,

    @SerialName("permit2_batch") Permit2Batch,
}

@Serializable
enum class GuardAmountError {
    @SerialName("invalid_amount") InvalidAmount,

    @SerialName("unlimited_disabled") UnlimitedDisabled,
}

@Serializable
enum class GuardBlockReason {
    @SerialName("off_chain_permit") OffChainPermit,

    @SerialName("dai_permit_full_balance") DaiPermitFullBalance,
}

@Serializable
enum class GuardEditorMode {
    @SerialName("requested") Requested,

    @SerialName("balance") Balance,

    @SerialName("custom") Custom,

    @SerialName("revoke") Revoke,

    @SerialName("grant") Grant,
}

@Serializable
enum class GuardSurface {
    @SerialName("none") None,

    @SerialName("permit_sign") PermitSign,

    @SerialName("approval_editor") ApprovalEditor,

    @SerialName("batch") Batch,
}

@Serializable
sealed class GuardChoice {
    @Serializable
    @SerialName("amount")
    data class Amount(val amount_raw: String) : GuardChoice()

    @Serializable
    @SerialName("revoke")
    data object Revoke : GuardChoice()

    @Serializable
    @SerialName("grant")
    data object Grant : GuardChoice()
}

@Serializable
sealed class GuardLocus {
    @Serializable
    @SerialName("calldata_word")
    data class CalldataWord(val word_index: Int) : GuardLocus()

    @Serializable
    @SerialName("typed_path")
    data class TypedPath(val path: String) : GuardLocus()
}

@Serializable
data class GuardTokenMetaEntry(val token: String, val symbol: String, val decimals: Int)

@Serializable
data class GuardTokenMetaView(
    val symbol: String = "",
    val decimals: Int = 18,
    val verified: Boolean = false,
    val loading: Boolean = false,
)

@Serializable
data class GuardDetectedApproval(
    val kind: GuardApprovalKind,
    val token_address: String? = null,
    val spender: String,
    val amount_raw: String? = null,
    val amount_bits: Int? = null,
    val is_unbounded: Boolean = false,
    val is_boolean_grant: Boolean = false,
    val is_reducing: Boolean = false,
    val editable: Boolean = false,
    val block_reason: GuardBlockReason? = null,
    val deadline: String? = null,
    val locus: GuardLocus,
)

@Serializable
data class GuardEditorView(
    val mode: GuardEditorMode? = null,
    val custom_text: String = "",
    val error: GuardAmountError? = null,
    val choice: GuardChoice? = null,
    val display_amount_raw: String? = null,
    val requested_finite: Boolean = false,
    val has_balance_cap: Boolean = false,
    val balance_raw: String? = null,
)

@Serializable
data class GuardIncreaseTotalView(val current: String? = null, val increment: String, val total: String? = null)

@Serializable
data class GuardLegView(
    val to: String,
    val approval: GuardDetectedApproval? = null,
    val meta: GuardTokenMetaView = GuardTokenMetaView(),
    val editor: GuardEditorView? = null,
    val choice: GuardChoice? = null,
    val needs_editor: Boolean = false,
    val needs_choice: Boolean = false,
    val grants_broad: Boolean = false,
)

@Serializable
data class GuardBatchView(
    val legs: List<GuardLegView> = emptyList(),
    val any_uncapped: Boolean = false,
    val any_to_own_token: Boolean = false,
    val all_settled: Boolean = false,
)

@Serializable
data class GuardView(
    val surface: GuardSurface = GuardSurface.None,
    val detected: GuardDetectedApproval? = null,
    val meta: GuardTokenMetaView = GuardTokenMetaView(),
    val editor: GuardEditorView? = null,
    val confirm_allowed: Boolean = true,
    val rewritten_params_json: String? = null,
    val increase_total: GuardIncreaseTotalView? = null,
    val decimals_unverified: Boolean = false,
    val expired: Boolean = false,
    val batch: GuardBatchView? = null,
)

@Serializable
sealed class GuardOperation {
    @Serializable
    @SerialName("read_token_metadata")
    data class ReadTokenMetadata(val chain_id: Int, val tokens: List<String>) : GuardOperation()

    @Serializable
    @SerialName("read_erc20_allowance")
    data class ReadErc20Allowance(val chain_id: Int, val token: String, val owner: String, val spender: String) : GuardOperation()

    @Serializable
    @SerialName("read_erc20_balance")
    data class ReadErc20Balance(val chain_id: Int, val token: String, val owner: String) : GuardOperation()
}

@Serializable
sealed class GuardShellResult {
    @Serializable
    @SerialName("meta_resolved")
    data class MetaResolved(val metas: List<GuardTokenMetaEntry>? = null) : GuardShellResult()

    @Serializable
    @SerialName("allowance_read")
    data class AllowanceRead(val allowance: String? = null) : GuardShellResult()

    @Serializable
    @SerialName("balance_read")
    data class BalanceRead(val balance: String? = null) : GuardShellResult()
}

@Serializable
sealed class GuardEvent {
    @Serializable
    @SerialName("approval_detected")
    data class ApprovalDetected(
        val method: String,
        val params_json: String,
        val chain_id: Int,
        val wallet_address: String? = null,
        val read_only: Boolean = false,
        val now_ms: Double,
    ) : GuardEvent()

    @Serializable
    @SerialName("preset_selected")
    data class PresetSelected(val mode: GuardEditorMode) : GuardEvent()

    @Serializable
    @SerialName("custom_amount_changed")
    data class CustomAmountChanged(val text: String) : GuardEvent()

    @Serializable
    @SerialName("grant_deliberately_chosen")
    data object GrantDeliberatelyChosen : GuardEvent()

    @Serializable
    @SerialName("revoke_chosen")
    data object RevokeChosen : GuardEvent()

    @Serializable
    @SerialName("leg_preset_selected")
    data class LegPresetSelected(val index: Int, val mode: GuardEditorMode) : GuardEvent()

    @Serializable
    @SerialName("leg_custom_amount_changed")
    data class LegCustomAmountChanged(val index: Int, val text: String) : GuardEvent()

    @Serializable
    @SerialName("leg_grant_deliberately_chosen")
    data class LegGrantDeliberatelyChosen(val index: Int) : GuardEvent()

    @Serializable
    @SerialName("leg_revoke_chosen")
    data class LegRevokeChosen(val index: Int) : GuardEvent()

    @Serializable
    @SerialName("batch_recipients_resolved")
    data class BatchRecipientsResolved(val recipients: List<List<String>>) : GuardEvent()
}

File: app-ios/VelaWallet/VelaWallet/Features/Signing/Core/GuardWire.swift
//
//  GuardWire.swift
//  VelaWallet
//
//  The `approval_guard` machine's view model, in Swift.
//
//  ## The mandate this machine exists to keep
//
//  An unlimited approval never leaves this wallet. Not as a warning somebody
//  can tap past — there is no "grant all anyway" chip, by the founder's own
//  ruling — but as a gate: `confirmAllowed` stays false until the person has
//  named a finite cap, and `rewrittenParamsJson` is what gets signed.
//
//  **`rewrittenParamsJson` being `nil` is not a green light.** It means
//  nothing was rewritten, *including when a rewrite failed*, and the untouched
//  params then meet the core's own refusal at the submit chokepoint. Failing
//  closed at two places is deliberate.
//
//  Amounts are DECIMAL STRINGS everywhere, never numbers. A `Double` cannot
//  hold 2^256 - 1, and rounding somebody's spending cap is the bug this whole
//  file exists to avoid.
//
//  Views are `Decodable` through `CoreJSON.decoder`; operations and results
//  stay dictionaries.
//

import Foundation

enum GuardSurface: String, Decodable {
    /// Not an approval — this machine has nothing to gate.
    case none
    /// An off-chain permit: signed verbatim under deliberate consent, never
    /// through the cap editor.
    case permitSign = "permit_sign"
    /// The editable never-unlimited spending-cap editor.
    case approvalEditor = "approval_editor"
    /// The EIP-5792 per-leg breakdown.
    case batch
}

enum GuardApprovalKind: String, Decodable {
    case erc20Approve = "erc20_approve"
    case increaseAllowance = "increase_allowance"
    case decreaseAllowance = "decrease_allowance"
    case setApprovalForAll = "set_approval_for_all"
    case erc2612Permit = "erc2612_permit"
    case daiPermit = "dai_permit"
    case permit2Single = "permit2_single"
    case permit2Batch = "permit2_batch"
}

/// Why editing is blocked. Semantic — the shell owns the words.
enum GuardBlockReason: String, Decodable {
    /// The dApp submits its own amount on-chain, so the wallet cannot cap it:
    /// rewriting would desync the signature and revert the dApp's own
    /// transaction. The honest advice is an on-chain approval instead.
    case offChainPermit = "off_chain_permit"
    /// A DAI permit grants full-balance access. Sign as requested, or reject.
    case daiPermitFullBalance = "dai_permit_full_balance"
}

enum GuardEditorMode: String, Decodable {
    case requested, balance, custom, revoke, grant
}

enum GuardAmountError: String, Decodable {
    case invalidAmount = "invalid_amount"
    case unlimitedDisabled = "unlimited_disabled"
}

/// Where the amount lives, for the rewrite step. The shell never uses this to
/// rewrite anything — the core does the re-encoding — but it decodes so the
/// view stays a faithful mirror.
enum GuardLocusWire: Decodable, Equatable {
    case calldataWord(index: Int)
    case typedPath(String)

    private enum Keys: String, CodingKey { case type, wordIndex, path }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: Keys.self)
        switch try container.decode(String.self, forKey: .type) {
        case "calldata_word":
            self = .calldataWord(index: try container.decode(Int.self, forKey: .wordIndex))
        case "typed_path":
            self = .typedPath(try container.decode(String.self, forKey: .path))
        case let other:
            throw DecodingError.dataCorruptedError(
                forKey: .type, in: container,
                debugDescription: "unknown guard locus '\(other)'"
            )
        }
    }
}

/// The person's decision for one approval.
enum GuardChoiceWire: Decodable, Equatable {
    /// A finite cap, raw base units as a decimal string.
    case amount(raw: String)
    /// 0 / false.
    case revoke
    /// Keep a boolean `true` — explicit and deliberate, and only ever for
    /// `setApprovalForAll` or a DAI permit.
    case grant

    private enum Keys: String, CodingKey { case type, amountRaw }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: Keys.self)
        switch try container.decode(String.self, forKey: .type) {
        case "amount": self = .amount(raw: try container.decode(String.self, forKey: .amountRaw))
        case "revoke": self = .revoke
        case "grant": self = .grant
        case let other:
            throw DecodingError.dataCorruptedError(
                forKey: .type, in: container,
                debugDescription: "unknown guard choice '\(other)'"
            )
        }
    }
}

struct GuardDetectedApprovalWire: Decodable, Equatable {
    let kind: GuardApprovalKind
    /// The ERC-20 token, NFT collection or Permit2 token, when known.
    let tokenAddress: String?
    /// Who is being granted spending power.
    let spender: String
    /// Raw base units, decimal string. `nil` for a boolean grant.
    let amountRaw: String?
    /// 256 or 160.
    let amountBits: Int?
    /// Effectively unlimited — at or above the cap, or a boolean grant-all.
    let isUnbounded: Bool
    /// A boolean grant: there is no amount to cap.
    let isBooleanGrant: Bool
    /// Reduces risk (a decrease or a revoke). Rendered as safe.
    let isReducing: Bool
    /// Whether a finite amount can be safely re-encoded for this shape.
    let editable: Bool
    let blockReason: GuardBlockReason?
    /// Unix seconds as a decimal string, when the shape carries a deadline.
    let deadline: String?
    let locus: GuardLocusWire
}

struct GuardTokenMetaViewWire: Decodable, Equatable {
    let symbol: String
    let decimals: Int
    /// The decimals came from the chain. When false the shell must say so —
    /// a cap read in the wrong decimals is off by orders of magnitude.
    let verified: Bool
    let loading: Bool
}

struct GuardEditorViewWire: Decodable, Equatable {
    /// `nil` only on the boolean card, before the deliberate tap.
    let mode: GuardEditorMode?
    let customText: String
    let error: GuardAmountError?
    /// `nil` ⇒ confirm stays disabled. That is the entire point of the
    /// machine.
    let choice: GuardChoiceWire?
    /// What the value row shows, raw base units.
    let displayAmountRaw: String?
    /// The "as requested" chip exists at all — a finite, non-zero incoming
    /// amount. When the request is unlimited the chip is drawn **disabled**,
    /// not merely unselected.
    let requestedFinite: Bool
    /// The one-tap finite balance cap is offered.
    let hasBalanceCap: Bool
    let balanceRaw: String?
}

/// The resulting-total row: "increase by 100" must never read as "cap at 100".
struct GuardIncreaseTotalViewWire: Decodable, Equatable {
    /// The on-chain allowance, when the read succeeded.
    let current: String?
    let increment: String
    /// `nil` means the read failed — and the row still warns that the
    /// increment ADDS to an existing allowance rather than hiding.
    let total: String?
}

struct GuardLegViewWire: Decodable, Equatable {
    let to: String
    let approval: GuardDetectedApprovalWire?
    let meta: GuardTokenMetaViewWire
    /// This leg's inline editor. The card renders it; it derives nothing.
    let editor: GuardEditorViewWire?
    let choice: GuardChoiceWire?
    let needsEditor: Bool
    /// This leg still blocks confirm.
    let needsChoice: Bool
    /// After the person's choice, this still grants broad access.
    let grantsBroad: Bool
}

struct GuardBatchViewWire: Decodable, Equatable {
    let legs: [GuardLegViewWire]
    let anyUncapped: Bool
    /// A leg that sends a token to its own contract burns it — the same
    /// fat-finger the single-send path flags, and far easier to miss buried
    /// in a batch.
    let anyToOwnToken: Bool
    let allSettled: Bool
}

struct GuardViewWire: Decodable, Equatable {
    let surface: GuardSurface
    let detected: GuardDetectedApprovalWire?
    let meta: GuardTokenMetaViewWire
    let editor: GuardEditorViewWire?
    /// **One of three gates.** False while an editable approval has no choice
    /// or a batch leg is unsettled. AND it with `sign_request`'s
    /// `confirmGateOpen` and `fee_policy`'s `confirmFeeReady`.
    let confirmAllowed: Bool
    /// The finite re-encode of the whole request, ready to submit. See the
    /// file header: `nil` is not a green light.
    let rewrittenParamsJson: String?
    let increaseTotal: GuardIncreaseTotalViewWire?
    /// Unverified decimals must be flagged explicitly on screen.
    let decimalsUnverified: Bool
    let expired: Bool
    let batch: GuardBatchViewWire?

    static let empty = GuardViewWire(
        surface: .none, detected: nil,
        meta: GuardTokenMetaViewWire(symbol: "", decimals: 18, verified: false, loading: false),
        editor: nil, confirmAllowed: true, rewrittenParamsJson: nil,
        increaseTotal: nil, decimalsUnverified: false, expired: false, batch: nil
    )
}
